rfcs/text/0014_notification_service.md

```diff
@@ -0,0 +1,373 @@
+
+The user can change the notification status to `read: false` manually in the Kibana UI.
+
+The user can change the status for all the existing notifications (`mark all as read`), even for those not-rendered in UI yet.
+
+Kibana HTTP API:
+- mark notification as read
+```
+POST /notification/read
+{ notification_id: string[] }
+```
+- Mark all the notification read
+```
+POST /notifications/_mark_all_read
+```
+- Pin a notification.
+```
+POST /notification/pin
+{ notification_id: string[] }
+```
+- Unpin a notification.
+```
+POST /notification/unpin
+{ notification_id: string[] }
+```
+
+Writing to the same Notification storage might cause conflict on the `write`. If the storage mechanism implementation
+doesn’t make them impossible, we will have to use a single place to change notification status (via the [queue](#localrepository)).
+
+## Settings
+Provides HTTP API for Kibana UI to configure user-specific settings.
+Not provided in MVP, we might add in the future:
+- Unsubscribe from a source
+- Subscribe to a source
+
+#### Filtering
+Filtering allows users to hide notifications they aren’t interested in. Kibana UI attaches a list of filters when requesting notifications.
+Kibana stores filter as a part of User settings data to provide a seamless experience between different deployments on Cloud and user sessions.
+
+
+# Drawbacks
+
+- It is not entirely clear what functionality will be supported by UNS and when this will be known.
+It forces us to evaluate each feature separately as to how difficult it will be to adapt it to UNS later.
+- The service relies on the yet-to-be-added ability to uniquely identify the user, enumerate users with a given set of permissions.
+- The service doesn't provide a built-in security mechanism but relies on a notification source implementation.
+- Lack of a centralized i18n service makes message transferring over-bloated due to the necessity to send all the possible translations over the wire.
+
+
+# Alternatives
+
+### Threat Cloud as another pluggable storage mechanism.
+Kibana notification service collects internal notifications and shows them in the UI for both on-prem and Cloud environments.
+Kibana reuses Notification service implementation on Cloud but enhances it with additional pluggable storage of external notifications.
+![img](../images/0014/alternative_implementation.png)
+
+The main benefit of this approach is that Kibana always uses the same service in any environment.
+But in addition to the restrictions listed in the [Drawbacks](#drawbacks) section, the following are added:
+- Each retrieval of a notification list still requires waiting for a roundtrip to UNS.
+- The list of Kibana notifications stored in an instance might be out of sync with a copy stored in UNS due to network problems. 
+- Storing notifications in the Cloud deployment and their copy in UNS consumes twice the disk space.
```

Comment:
I'd add that the pagination and sorting logic becomes more complicated as well.

Reply:
It shouldn't be a problem if we implement pagination via `search_after` and support sorting by the creation time only?
But it requires an additional roundtrip to UNS for every request, including navigation to the next page.

Reply:
Maybe I misunderstood. What I meant is that if we are polling 2 sources, the pagination logic becomes more problematic because to get the first 10 items, it requires us to retrieve 10 items from each source, and locally sort them and return the first 10 _global_ items.

Reply:
> Maybe I misunderstood. What I meant is that if we are polling 2 sources, the pagination logic becomes more problematic because to get the first 10 items, it requires us to retrieve 10 items from each source, and locally sort them and return the first 10 global items.

You are right, I merely didn't consider sorting as the main problem in this case, but waiting for a request is.